### PR TITLE
Compact rcf

### PR DESCRIPTION
**Pull request by sudiptoguha**
*Description of changes:* Refactors addPoint and readies the transition to float[].

### CODE REVIEW

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 1]
```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
...
+            if (parent == NULL) {
+                rootIndex = mergedNode;
+            } else {
+                replaceNode(siblingOffset, mergedNode);
+            }
+            leafNodes.parentIndex[leafOffset - maxSize] = mergedNode;
+            if (isLeaf(siblingOffset))
+                leafNodes.parentIndex[siblingOffset - maxSize] = mergedNode;
+            else
+                internalNodes.parentIndex[siblingOffset] = mergedNode;
```
Comment (wnbts) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:269]: parentheses are missing

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 2]
```diff
@@ -327,138 +421,75 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
...
+     * <li>If the point is within the current bounding box then make the change
+     * corresponding to the most recent location and "resolve" the situation. If the
+     * point is outside the bounding box of the tree then this step is taken at the
+     * root node in addpoint(Sequential seg)</li>
      * </ol>
      *
      * @param nodeOffset the current node in the tree we are on
      * @param point      the point that we want to add to the tree
      * @param pointIndex is the location of the point in pointstore
      */
```
Comment (wnbts) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:445]: the return of this method is missing

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 3]
```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
...
+                siblingOffset = nodeOffset;
+                savedBox = mergedBox;
+            }
+            return false;
+        }
+    }
+
+    CandidateChange candidateChange = new CandidateChange();
+
+    boolean checkEquals(double[] point, double[] otherPoint) {
```
Comment (wnbts) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:310]: minor. inputs should be sanitized and array comparison can be done by Arrays.equals without an implementation everywhere it is needed.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 4]
```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
...
+                cutValue = splitValue;
+                cutDimension = splitDimension;
+                siblingOffset = nodeOffset;
+                savedBox = mergedBox;
+            }
+            return false;
+        }
+    }
+
+    CandidateChange candidateChange = new CandidateChange();
```
Comment (wnbts) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:308]: the used data will not be garbage collected. is that desired?

**Review by wnbts — COMMENTED** [review]
documentation, interface, and testing can use some work

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 4 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:308]: Yes. This structure is one per tree and simply facilitates the addPoint. We simply overwrite the data.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 3 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:310]: Agree. This is a discussion point - we now have equals in the PointStore. But the issue is that the equals in the addPoint has to be in sync with the randomCut function. Maybe we remove the pointEquals in PointStore?

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 2 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:445]: Yes! Fixed.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 1 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:269]: fixed.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 5]
```diff
@@ -327,138 +421,75 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
...
-        int pointIndex = seq.getValue();
-        double[] pointValue = pointStore.get(pointIndex);
-        if (rootIndex == NULL) {
-            rootIndex = (short) (leafNodes.add(NULL, pointIndex, 1) + maxSize);
-            return pointIndex;
-        } else {
-            return addPoint(rootIndex, pointValue, pointIndex);
-        }
-    }
+    boolean addResolved;
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:424]: done

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 6]
```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
...
+        return (point[dimension] <= val);
+    }
+
+    class CandidateChange {
+        short siblingOffset;
+        int cutDimension;
+        double cutValue;
+        BoundingBox savedBox;
+
+        void setCandidateChange(short sibling, short dim, double val, BoundingBox box) {
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:237]: The point is to simultaneously store all the information -- that is why it made sense as an inner class. These three fields should correspond to the same act of "cut", and should not be changed one at a time. We can break this up, but not may be good to keep the conceptual link.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 7]
```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
...
+
+    boolean leftOf(double[] point, int dimension, double val) {
+        return (point[dimension] <= val);
+    }
+
+    class CandidateChange {
+        short siblingOffset;
+        int cutDimension;
+        double cutValue;
+        BoundingBox savedBox;
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:235]: True, fixed.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 8]
```diff
@@ -220,7 +220,100 @@ protected boolean leftOf(float[] point, short nodeOffset) {
     }
 
     protected boolean leftOf(double[] point, short nodeOffset) {
-        return (point[internalNodes.cutDimension[nodeOffset]] <= internalNodes.cutValue[nodeOffset]);
+
+        return leftOf(point, internalNodes.cutDimension[nodeOffset], internalNodes.cutValue[nodeOffset]);
+    }
+
+    boolean leftOf(double[] point, int dimension, double val) {
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:227]: fixed.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/BoundingBox.java [hunk 9]
```diff
@@ -123,6 +123,34 @@ public BoundingBox getMergedBox(double[] point) {
...
+        }
+        return new BoundingBox(minValuesMerged, maxValuesMerged);
+    }
+
+    public BoundingBox addPoint(double[] point) {
+        rangeSum = 0;
+        for (int i = 0; i < point.length; ++i) {
+            minValues[i] = Math.min(minValues[i], point[i]);
+            maxValues[i] = Math.max(maxValues[i], point[i]);
+            rangeSum += maxValues[i] - minValues[i];
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/BoundingBox.java:149]: rangeSum is used in the randomCut function in both the original and compact RCF.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 10]
```diff
@@ -327,138 +351,176 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
...
+        void setCandidateChange(short sibling, short dim, double val, BoundingBox box) {
+            siblingOffset = sibling;
+            cutDimension = dim;
+            cutValue = val;
+            savedBox = box;
         }
 
-        // either the node is not a leaf, or else it's a leaf node containing a
-        // different point
+        void effectChange(int pointIndex, double[] point, int parentIndex) {
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:384]: We could call this method `resolve` to match the flag name.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 11]
```diff
@@ -327,138 +351,176 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
...
+     * information of the parent where the addition of the point no longer updates
+     * any bounding box.
      */
 
-    private int addPoint(short nodeOffset, double[] point, int pointIndex) {
-        if (isLeaf(nodeOffset) && pointStore.pointEquals(leafNodes.pointIndex[nodeOffset - maxSize], point)) {
-            // the inserted point is equal to an existing leaf point
-            ++leafNodes.mass[nodeOffset - maxSize];
-            return leafNodes.pointIndex[nodeOffset - maxSize];
+    class CandidateChange {
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:371]: Will this class be reused in the future for deletes? If not, maybe we could come up with a better name than "change".

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 12]
```diff
@@ -327,138 +351,176 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
...
+        short siblingOffset;
+        int cutDimension;
+        double cutValue;
+        BoundingBox savedBox;
+
+        void setCandidateChange(short sibling, short dim, double val, BoundingBox box) {
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:377]: suggestion: call this `initialize` or `init`.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 13]
```diff
@@ -327,138 +351,176 @@ private int deletePoint(short nodeOffset, double[] point, int level) {
...
+            addResolved = true;
+            int oldmass = (isLeaf(siblingOffset)) ? leafNodes.mass[siblingOffset - maxSize]
+                    : internalNodes.mass[siblingOffset];
+            short leafOffset = (short) (leafNodes.add(NULL, pointIndex, 1) + maxSize);
+            short mergedNode = leftOf(point, cutDimension, cutValue)
+                    ? internalNodes.addNode(NULL, leafOffset, siblingOffset, cutDimension, cutValue, oldmass + 1)
+                    : internalNodes.addNode(NULL, siblingOffset, leafOffset, cutDimension, cutValue, oldmass + 1);
+
+            int parent;
+            if (isLeaf(siblingOffset))
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:394]: minor: braces around single line statements.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 14]
```diff
@@ -220,17 +223,39 @@ protected boolean leftOf(float[] point, short nodeOffset) {
     }
 
     protected boolean leftOf(double[] point, short nodeOffset) {
-        return (point[internalNodes.cutDimension[nodeOffset]] <= internalNodes.cutValue[nodeOffset]);
+        return leftOf(point, internalNodes.cutDimension[nodeOffset], internalNodes.cutValue[nodeOffset]);
     }
 
-    private boolean resolvedDelete;
+    private boolean leftOf(double[] point, int dimension, double val) {
+        return (point[dimension] <= val);
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:230]: fixed

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java [hunk 15]
```diff
@@ -220,17 +223,39 @@ protected boolean leftOf(float[] point, short nodeOffset) {
...
+    /**
+     * Checks equality of points -- this has to work in tandem across deletePoint,
+     * addPoint and randomCut
+     * 
+     * @param point      first point in question
+     * @param otherPoint second point
+     * @return identical or otherwise
+     */
+
+    private boolean checkEquals(double[] point, double[] otherPoint) {
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:242]: This brings in the same comparison used in PointStore. I think we need to still perform the cleanCopy for updates. We can update this later.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 11 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:371]: Fair. Using CanadidateForSiblingInAddPoint, which is the exact purpose. It is unlikely to be useful in the current delete method (but it is not impossible that something like this becomes necessary for deletes once more information is added to the nodes in the trees)

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 12 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:377]: Fixed.

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 10 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:384]: Fixed

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java (hunk 13 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeDouble.java:394]: fixed

**Comment by jotok** [comment]
Merging now, since there seems to be an issue with the GitHub workflow and we've confirmed that tests pass offline.

**Pull request merged**